docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,167 @@
+FFI, and for operational reasons an RPC mechanism makes more sense.
+
+The socket protocol has not changed substantially since its original design,
+and has the advantage of being simple to implement in almost any reasonable
+language.  However, its simplicity includes some limitations that have had a
+negative impact on the stability and performance of out-of-process applications
+using it. In particular:
+
+- The protocol lacks request identifiers, so the client and server must
+  complete requests in strict FIFO order. Even if the client issues requests
+  that have no dependency on each other, the protocol has no way except order
+  of issue to map responses to requests. This strictly limits the application's
+  ability to process requests concurrently, and has been a source of complaints
+  from some network operators on that basis.
+
+- The protocol lacks method identifiers, so the only way for the client and
+  server to understand which operation is requested is to dispatch on the type
+  of the request and response payloads. For responses, this means that any
+  error condition is terminal not only to the request, but to the entire ABCI
+  connection.
+
+  The intent historically seems to have been that all ABCI errors are protocol
+  fatal. In practice, however, this greatly complicates debugging a faulty
+  node, since the only way to respond to errors it to panic the node which
+  loses valuable context that could have been logged.
+
+- There are subtle concurrency management dependencies between the client and
+  the server that are not clearly documented anywhere, and it is very easy for
+  small changes in both the client and the server to lead to tricky deadlocks,
+  panics, race conditions, and slowdowns. As a recent example of this, see
+  https://github.com/tendermint/tendermint/pull/8581.
+
+These limitations are fixable, but one important question is whether it is
+worthwhile to fix them.  The stability and language coverage of gRPC have
+improved a lot, and today it is probably simpler to set up and maintain an
+application using gRPC transport than to reimplement the Tendermint socket
+protocol.
+
+Moreover, gRPC addresses all the above issues out-of-the-box, and requires
+(much) less custom code for both the server (i.e., the application) and the
+client. The project is well-funded and widely-used, which makes it a safe bet
+for a dependency.
+
+## Decision
+
+There is a set of related alternatives to consider:
+
+- Question 1: Designate a single IPC standard for out-of-process applications?
+
+  Claim: We should converge on one (and only one) IPC option for out-of-process
+  applications. We should choose an option that, after a suitable period of
+  deprecation, can address most or all the highest-impact uses of Tendermint.
+  Maintaining multiple options increases the surface area for bugs and
+  vulnerabilities, and we should not have multiple options for basic interfaces
+  without a clear and well-documented reason.
```

Comment:
I personally see very little issue with this, as long as, as you say, the highest-impact uses of Tendermint can be covered. Maintaining two clients with very overlapping sets of functionality is quite costly and results in a worse user experience for both as a result.